### DISCUSSION TITLE
Using axum-sessions with axum-test-helper

### DISCUSSION THREAD

**Discussion by aaronleopold** [maxcountryman/axum-sessions#33] [Q&A]
Hello!

I wanted to try and start unit testing a server I've been developing but haven't quite figured out the best way to go about mocking the session handles throughout my router handlers.

A simple example might be:

```rust
async fn viewer(session: ReadableSession) -> ApiResult<Json<User>> {
	if let Some(user) = session.get::<User>("user") {
		Ok(Json(user))
	} else {
		Err(ApiError::Unauthorized)
	}
}

#[cfg(test)]
mod tests {
	use axum_test_helper::TestClient;

	use crate::{
		config::state::{AppState, AppStateMock},
		routers::get_test_router,
	};

	#[tokio::test]
	async fn viewer_with_valid_session() {
		let mock_state = AppState::_mock();
		let router = get_test_router(mock_state.clone());
		let app = TestClient::new(router);
		// FIXME: this is private! I need to figure out how to fake a session...
		// let session_handle = session_layer.load_or_create(cookie_value.clone()).await;
		todo!()
	}
}
```

I think my ideal would be if it was possible to do something along the lines of layering some sort of `SessionLayerMock` to my app router, rather than a real `SessionLayer`, that gives me access to do the following before a request: 

```rust
session_layer.expect_get<User>(User {..})
```

**Comment by maxcountryman** [comment]
We've setup some basic integration tests with [tower-sessions](https://github.com/maxcountryman/tower-sessions). They might give you some idea of one possible approach.
